docs/design/proposals/healer-controller.md

```diff
@@ -0,0 +1,120 @@
+* Moving the k8s specific code out of ceph-csi, to make ceph-csi orchestrator independent.
+* Achieve a way to monitor the health of NBD volumes.
+* We currently do not have a way to act on failed/crashed NBD volumes with the existing healer logic. Once the controller detects that the volume is not healthy, it will have an opportunity to act on it.
+* This will be a generic framework for any userspace mounters added in the future, example ceph-fuse
+
+# Details
+
+                    .-------------------.
+                    |     CSI-Addons    |
+                    | Healer Controller |
+                    '-------------------'
+                              |
+                              | gRPC
+                              |
+            .---------+------------------------------.
+            |         |                              |
+            |  .------------.        .------------.  |
+            |  | CSI-Addons |  gRPC  |    CSI     |  |
+            |  |  side-car  |--------| NodePlugin |  |
+            |  '------------'        '------------'  |
+            | CSI-driver Pod                         |
+            '----------------------------------------'
+
+## CSI-Addons Healer Controller
+* The healer controller will be implemented as part of the CSI-Addons controller driver. It is the one who triggers controller RPC to check the health condition of the CSI volumes.
+* Watch for the `CSIAddonsNode` object, if the object is created/modified by the side-car do below operations
+    * Fetch all the volume attachments (VA) list
+    * Filter the volume attachments list through matching driver name and status attached
+    * For each volume attachment get the respective PV information and check the criteria of PV Bound (also we can have configuration options like mounter type)
+    * Send the RPC request for side-car to check if the volume is healthy along with the PV details details
+    * Log the final response returned from the side-car
+* The healer controller periodically triggers the health status checks of the CSI volumes via the gRPC calls.
+
+## CSI-Addons side-car
+* The side-car registers itself by creating a `CSIAddonsNode` CR that the CSI-Addons Controller can use to connect to the side-car and execute operations.
+* When a nodeplugin container is restarted the side-car will update/create(if not present) [CSIAddonsNode](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/api/v1alpha1/csiaddonsnode_types.go) object.
+* The side-car works as relay, upon receiving a gRPC call from healer controller, the CSI-Addons side-car will then check volume's mounting conditions collected via the CSI NodeGetVolumeStats RPC.
+* In case if any volume condition is reported abnormal, the CSI-Addons side-car will act on it, it does some internal operation like fetching secrets, constructing staging path and finally sends NodeStageVolume request to the CSI driver and finally returns back the response to the controller.
```

Comment:
Just sent a heal call to csidriver and let it decide whether node stage req is required or not.

Reply:
I know it looks simple but, I think this needs more cycles for fetching secrets, formatting the staging path, and capturing/collecting more details needed to make the NodeStageCall(), for almost all the volume even if they are healthy.

Reply:
> I know it looks simple but, I think this needs more cycles for fetching secrets, formatting the staging path, and capturing/collecting more details needed to make the NodeStageCall(), for almost all the volume even if they are healthy.

sidecar is just a relay, 
its better to have more api calls in single request flow from controller to sidecar to csidriver
than to have back and forth between sidecar and csi-driver while controller sits waiting.

Reply:
I agree that this is a bit debatable topic on its own.
1. For restart of the nodeplugin case, it will mostly need NodeStageVolume() for 100% of PVs
2. For periodic checks, I'm optimistic that most of the time the health of the volume is in a good state, only the corner cases are process restarts (considering userspace mounters). Hence I think sending NodeStageVolume() all the time is a bit costly.

So maybe a hybrid model is needed to satisfy and save some jiffies here, like
* For restart case send a NodeStageVolume()
* For periodic health checks scenarios, send a NodeGetVolumeStats() first and then send a NodeStageVolume() if the status of the PV is set to abnormal?

Reply:
CSI-Addons can not send CSI commands, those are reserved for the CO.

CSI-drivers can decide on their own implementation of healing, maybe they want to call their own NodeStageVolume procedure. Don't mix standard CSI procedures with CSI-Addons procedures.

There is a difference between regular health checks, and an informational message "you restarted, and you should make sure the volume is still available". The last part is expected when a CSIAddonsNode is detected, the 1st is for https://github.com/kubernetes-csi/external-health-monitor/

Reply:
For now, as we are focusing on restarting node plugin case, we might want to fill in all details in one shot and send. As discussed for periodic healing we might need some kind of a hybrid model to reduce the cycles involved in collecting details which might be unneeded in most of the times

Reply:
> CSI-Addons can not send CSI commands, those are reserved for the CO.

@nixpanic We wanted to send NodeGetVolumeStats() to check the volume condition, why not NodeStageVolume() ?

Reply:
I was thinking we can implement a function like HealVolume() that can be called from sidecar which does NodeGetVolumeStats() and  Heal() in one go, as @Rakshith-R pointed above

Reply:
A function like `HealVolume()` would need implementation in the CSI-driver (called from the sidecar). CSI-drivers needs to handle healing differently, I expect.

`NodeGetVolumeStats()` is a CSI call, so can not be called from CSI-Addons operations. You need to design an operation that the CSI-Addons sidecar can call on the CSI-driver. The design for `Heal()` and a `HealthCheck()` function need to go to csi-addons/spec.

Reply:
> A function like HealVolume() would need implementation in the CSI-driver (called from the sidecar). CSI-drivers needs to handle healing differently, I expect.

Yes

>NodeGetVolumeStats() is a CSI call, so can not be called from CSI-Addons operations. You need to design an operation that the CSI-Addons sidecar can call on the CSI-driver. The design for Heal() and a HealthCheck() function need to go to csi-addons/spec.

@nixpanic 
 oh! sorry if that confused you, I now understand why we do not want to use NodeGetVolumeStats() served by the CSI driver which was for CSI Spec. I mean NodeGetVolumeStats() RPC is also served by CSI-addons server. Do we see any reason why we do not want to use the same RPC signature logically? do you think it just creates some additional confusion? if so we can just name it NodeGetVolumeStatus().

Reply:
Or maybe I will just have the new RPC naming it as NodeGetVolumeStatus() which is mostly a copy of NodeGetVolumeStats()

Reply:
> The design for Heal() and a HealthCheck() function need to go to csi-addons/spec.

@nixpanic I expect I will have more clarity on Heal() and a HealthCheck() functions while I work on this RFE, so I hope to start on the draft implementation first and then send a PR to csi-addons/spec, only after the spec gets merged I will send the actual controller PR.

Hope that won't hurt.